### PR TITLE
Extend world generation

### PR DESCRIPTION
**Pull request by vallsv**
Hi again, here is my current work.

My aim was to play with underground caves. But for that it was needed to do many things.

Chunk was needed for example. So i have reworked the model (inside a new `world` module). It is named `Section`, as it was named like that in the base code. It is now an independent class where blocks are stored using `SECTION_SIZE^3`. Now it also have a fixed size vertically. The `Model` do not deal any more with single blocks (except the methods to add and remove blocks). And the rendering is done a single time per section (there is a single vertex list per section). I feel like it reduces a lot the complexity.

The world generator can now deal with vertical sections and with any section size. So it is easy to play with the underground generation, or with the trees. There is a new `noise` module to deal with octave params.

Finally there is new blocks for the stones and the ores. So the texture was updated with a bigger one to have more slots, and i have flipped the texture binding (bottom-up -> top-down), i think it is much more convenient.

Hope that's not too much, else you anyway can pick what's you like.

To test the result you can switch the `enclosure` to false and the world became infinite (on the sides and under your shoes).

```
-        self.enclosure = True
+        self.enclosure = False
```

Capture d’écran de 2020-05-21 20-53-00

### CODE REVIEW

**Comment by vallsv** [comment]
BTW, as there is no shading it's difficult to distinguish anything underground. That's not really interesting for play there.

File: game/world.py [hunk 1]
```diff
@@ -262,6 +262,9 @@ def on_sector_received(self, chunk):
         to the main thread.
         """
         self._enqueue(self.register_sector, chunk)
+        # This sleep looks to be needed to reduce the load of the main thread.
+        # Maybe it also release the GIL and reduce the coupling with the main thread.
+        time.sleep(0.01)
```
Comment (benmoran56) [inline game/world.py:267]: I think your hunch is correct. Good catch!

**Comment by benmoran56** [comment]
I've gone though your changes and everything looks OK to me. Actually, "infinite world" was one of the things that was being thought about already. 

Regarding the textures, I was planning to switch them to individual images, and using the `pyglet.resource` module to load them (which will automatically place them in a Texture Atlas). This can be done any time, but the idea is that it's easier to update/add textures later. 

I'll clone your repo later today and give this a further review. Thanks for all of your contributions!

**Comment by benmoran56** [comment]
One other comment:  I was thinking that it's best to avoid Shaders while we're still using the current pyglet release. Once pyglet v2.0 is released in the (hopefully near) future, modern Shaders would be easier to do and better understood by others.  In the mean time, I was considering making simple light classes using legacy OpenGL calls.

**Pull request merged**